### DISCUSSION TITLE
Cannot connect to db from WSL1, "[Errno 22] Invalid argument"

### DISCUSSION THREAD

**Discussion by edvin-svk** [oracle/python-oracledb#69] [Q&A]
I'm having problems connecting to a db from WSL1. Is this supposed to be possible? I'm on python 3.8.6 with oracledb 1.1.1. When doing
`oracledb.connect(user=un, password=pw, dsn='myhost/myservice')`
I get
```
File ".../oracledb/connection.py", line 1015, in connect
 File ".../oracledb/connection.py", line 136, in __init__
    else:
  File "src/oracledb/impl/thin/connection.pyx", line 298, in oracledb.thin_impl.ThinConnImpl.connect
  File "src/oracledb/impl/thin/connection.pyx", line 189, in oracledb.thin_impl.ThinConnImpl._connect_with_params
  File "src/oracledb/impl/thin/connection.pyx", line 162, in oracledb.thin_impl.ThinConnImpl._connect_with_description
  File "src/oracledb/impl/thin/connection.pyx", line 116, in oracledb.thin_impl.ThinConnImpl._connect_with_address
  File "src/oracledb/impl/thin/protocol.pyx", line 213, in oracledb.thin_impl.Protocol._connect_phase_two
OSError: [Errno 22] Invalid argument
```
Any idea what I might be doing wrong?

**Comment by cjbj** [comment]
Can you describe what you've installed and how?  What DB version are you using?   Is your DSN as simple as you show it or are you using any special options?   Does Thick mode connect?

**Reply by edvin-svk** [reply]
What I installed is simply oracledb using poetry. Did you have something else in mind?

The db is "Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production Version 19.13.0.0.0".

The DSN I can successfully connect with (via PyCharm) is on the form 
jdbc:oracle:thin:@( DESCRIPTION = (ADDRESS_LIST = (LOAD_BALANCE=off) (FAILOVER=ON) (ADDRESS=(PROTOCOL=TCP)(HOST= myhost)(PORT=1521))) (CONNECT_DATA=(SERVICE_NAME= myservicename)))

I don't know how to try thick mode, can you help with that?
EDIT: sorry ready the installation instructions more carefully now. I didn't yet install the client libraries so calling `init_oracle_client()` returns error about .so file not found. Will try to install these

**Reply by anthony-tuininga** [reply]
The DSN you supplied is for Java, not Python! Strip off the 'jdbc:oracle:thin:@' from the front of the DSN and try that instead. You can also use this instead:

```python
conn = oracledb.connect(user=USER, password=PASSWORD, dsn="myhost:1521/myservicename")
```

**Reply by cjbj** [reply]
Maybe try disabling out-of-band breaks by using `conn = oracledb.connect(user=USER, password=PASSWORD, dsn="myhost:1521/myservicename", disable_oob=True)`.  If my reading of the code is correct this would skip the particular line of code throwing the error.  @anthony-tuininga may know differently.  Of course we'd like to be able solve the base problem, but this may help.

**Reply by anthony-tuininga** [reply]
That is correct. No attempt is made to send out-of-band data if `disable_oob` is set to the value `True`. I did some searching but I was unable to find a way to determine if OOB is supported. Perhaps just catching this error and disabling OOB may be suffiicent?

**Reply by edvin-svk** [reply]
> The DSN you supplied is for Java, not Python! Strip off the 'jdbc:oracle:thin:@' from the front of the DSN and try that instead. You can also use this instead:
> 
> ```python
> conn = oracledb.connect(user=USER, password=PASSWORD, dsn="myhost:1521/myservicename")
> ```

Yes I know, I just wanted to show in which way I managed to connect successfully in case the special options there are of any importance. In python I use what you wrote, as I explained in my first post.

> Maybe try disabling out-of-band breaks by using `conn = oracledb.connect(user=USER, password=PASSWORD, dsn="myhost:1521/myservicename", disable_oob=True)`. If my reading of the code is correct this would skip the particular line of code throwing the error. @anthony-tuininga may know differently. Of course we'd like to be able solve the base problem, but this may help.

Yes great, with this option it works, thank you!
As additional information I can add that I can connect successfully with `disable_oob=False` using thick mode (i.e. installing the client libraries and calling `init_oracle_client()`).

**Comment by anthony-tuininga** [comment]
Great! Can you try putting the try/except block around the offending lines, like this?

```python
        # if we can use OOB, send an urgent message now followed by a reset
        # marker to see if the server understands it
        if self._caps.supports_oob \
                and self._caps.protocol_version >= TNS_VERSION_MIN_OOB_CHECK:
            try:
                self._socket.send(b"!", socket.MSG_OOB)
                self._send_marker(self._write_buf, TNS_MARKER_TYPE_RESET)
            except OSError:
                self._caps.supports_oob = False
```

**Reply by edvin-svk** [reply]
Hm that's odd, I don't have the `impl` directory in the oracledb directory. Instead I have .so files for the thin_impl, thick_impl, base_impl modules. So I can't access the code. Is this the expected layout?

**Reply by anthony-tuininga** [reply]
Yes, for a binary installation! You will have to grab the source code and build it. The instructions are found [here](https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html#installing-from-source-code).
